### PR TITLE
Electron Go: restart the tun2socks process when it crashes

### PR DESCRIPTION
**Pull request by alalamav**
- Restarts the Go tun2socks process when it exits without the user disconnecting.
- Windows: preemptively stops the tun2socks process on suspend, so as not to trigger the restart logic.

### CODE REVIEW

File: src/electron/go_vpn_tunnel.ts [hunk 1]
```diff
@@ -136,10 +136,12 @@ export class GoVpnTunnel implements VpnTunnel {
   }
 
   private suspendListener() {
-    // Swap out the current listener, restart once the system resumes.
+    // Windows: when the system suspends, tun2socks terminates due to the TAP device getting closed.
+    // Preemptively stop tun2socks without notifying its exit.
     this.tun2socks.onExit = () => {
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:141]: Remove this. I think changing this handler is confusing. The output in the stop() method can differentiate stop from crash. Perhaps you just have a log output here saying it's suspending.

File: src/electron/go_vpn_tunnel.ts [hunk 2]
```diff
@@ -262,8 +265,15 @@ class GoTun2socks {
           return;
         }
         console.debug('tun2socks started');
-        this.process.onExit = (code?: number, signal?: string) => {
+        this.isRunning = true;
+        this.process.onExit = async (code?: number, signal?: string) => {
           console.debug('tun2socks stopped');
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:270]: Output this only if `!this.isRunning`

File: src/electron/go_vpn_tunnel.ts [hunk 3]
```diff
@@ -262,8 +265,15 @@ class GoTun2socks {
...
+          if (this.isRunning) {
+            // The process crashed, restart it without notifying its exit.
+            console.warn(`tun2socks crashed, restarting...`);
```
[the remark this answers is not in the document]
Comment (fortuna) [reply src/electron/go_vpn_tunnel.ts:273]: Also, this may trigger if the process exits without crashing for some reason. Perhaps we should say "unexpectedly exited" instead of "crashed".

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/electron/go_vpn_tunnel.ts (hunk 1 above)
Comment (alalamav) [reply src/electron/go_vpn_tunnel.ts:141]: I agree that swapping the listener is not the best. The issue with that approach is that the process gets restarted immediately after receiving the suspend event, at which point the TAP device is not available, so restarting fails.

File: src/electron/go_vpn_tunnel.ts [hunk 4]
```diff
@@ -136,10 +136,11 @@ export class GoVpnTunnel implements VpnTunnel {
   }
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:136]: (code above) I don't think most of that logic in the GoVpnTunnel constructor is needed. Tun2Socks will never exit unless we explicitly call stop(). You should be able to get rid of `this.tun2socks.onExit = this.tun2socksExitListener;`

Also, that entire setup should happen in `connect()`, not the constructor.

**Review by fortuna — CHANGES_REQUESTED** [review]
It's hard to follow the onExit logic. I believe we can set it once and never change it. See my suggestions.

File: src/electron/go_vpn_tunnel.ts (hunk 1 above)
Comment (alalamav) [reply src/electron/go_vpn_tunnel.ts]: Done.

File: src/electron/go_vpn_tunnel.ts [hunk 5]
```diff
@@ -272,11 +281,13 @@ class GoTun2socks {
         resolve();
       };
       this.process.launch(args);
+      this.isRunning = true;
     });
   }
 
   stop() {
```
[the remark this answers is not in the document]
Comment (alalamav) [reply src/electron/go_vpn_tunnel.ts]: Done - `stop` now return a promise.

File: src/electron/go_vpn_tunnel.ts (hunk 4 above)
Comment (alalamav) [reply src/electron/go_vpn_tunnel.ts:126]: Done. I moved some of the logic to `connect`; however, to remain compatible with the `VpnTunnel` this class still needs to support the `onceDisconnected` promise.

**Review by alalamav — COMMENTED** [review]
Thank you for the suggestions. 

I removed the tun2socks exit listener and`stop` now returns a promise. I also removed the `isRunning` variable, since we're restarting tun2socks unconditionally `onExit` and swap the handler when `stop` is called.

File: src/electron/go_vpn_tunnel.ts [hunk 6]
```diff
@@ -262,7 +264,14 @@ class GoTun2socks {
           return;
         }
         console.debug('tun2socks started');
-        this.process.onExit = (code?: number, signal?: string) => {
+        this.process.onExit = async (code?: number, signal?: string) => {
```
[the remark this answers is not in the document]
Comment (fortuna) [reply src/electron/go_vpn_tunnel.ts:255]: `spawn()`, called by `launch()` will throw on invalid binary: 
image

You should catch the launch call and reject in case of errors. I think that would be cover everything?

File: src/electron/go_vpn_tunnel.ts [hunk 7]
```diff
@@ -275,12 +265,14 @@ class GoTun2socks {
...
-  stop() {
-    this.process.stop();
-  }
-
-  set onExit(listener: ((code?: number, signal?: string) => void)|undefined) {
-    this.exitListener = listener;
+  async stop() {
+    return new Promise<void>((resolve) => {
+      this.process.onExit = (code?: number, signal?: string) => {
+        console.debug('tun2socks stopped');
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:271]: Output the code and signal.

File: src/electron/go_vpn_tunnel.ts [hunk 8]
```diff
@@ -150,8 +139,6 @@ export class GoVpnTunnel implements VpnTunnel {
     }
 
     console.log('restarting tun2socks after resume');
-
-    this.tun2socks.onExit = this.tun2socksExitListener;
     this.tun2socks.start(this.isUdpEnabled);
 
     // Check if UDP support has changed; if so, silently restart.
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:144]: Could you rename `retestUdp` to `updateUdpSupport` or something like that. I find the restest name misleading because it makes me think it's just testing something and not changing any state.

File: src/electron/go_vpn_tunnel.ts [hunk 9]
```diff
@@ -172,25 +159,27 @@ export class GoVpnTunnel implements VpnTunnel {
...
+    // Restart tun2socks.
+    await this.tun2socks.stop();
+    await this.tun2socks.start(this.isUdpEnabled);
   }
 
   // Use #onceDisconnected to be notified when the tunnel terminates.
   async disconnect() {
-    powerMonitor.removeListener('suspend', this.suspendListener.bind(this));
-    powerMonitor.removeListener('resume', this.resumeListener.bind(this));
+    if (this.terminated) {
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:169]: Rename to disconnected, so the vocabulary is consistent

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/electron/go_vpn_tunnel.ts [hunk 10]
```diff
@@ -199,9 +188,11 @@ export class GoVpnTunnel implements VpnTunnel {
       // connected.
       console.error(`could not stop routing: ${e.message}`);
     }
+    this.resolveAllHelpersStopped();
```
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:191]: Shouldn't we communicate to the user if disconnection fails?
Perhaps we should reject the promise in case of failure?

**Review by fortuna — APPROVED** [review]

File: src/electron/go_vpn_tunnel.ts (hunk 10 above)
Comment (alalamav) [reply src/electron/go_vpn_tunnel.ts:191]: Good point. One operation shouldn't fail the call, but we can keep track of the errors and throw if both fail.
